pkg/identity/identity.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2019, Oath Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package identity
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/yahoo/athenz/clients/go/zms"
+	"github.com/yahoo/athenz/libs/go/zmssvctoken"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/crypto"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/log"
+)
+
+// PrivateKeyProvider returns a signing key on demand.
+type PrivateKeyProvider func() (*crypto.SigningKey, error)
+
+// tokenProvider provides tokens.
+type tokenProvider interface {
+	Token() (string, error)
+}
+
+// TokenProvider implements tokenProvider provides nTokens
+type TokenProvider struct {
+	config  Config
+	l       sync.RWMutex
+	current string
+	expire  time.Time
+}
+
+// Config is the token provider configuration.
+type Config struct {
+	Client             *zms.ZMSClient
+	Header             string
+	Domain             string             // Athenz domain
+	Service            string             // Athenz service
+	PrivateKeyProvider PrivateKeyProvider // source for private keys
+	TokenExpiry        time.Duration      // token expire time
+}
+
+// NewTokenProvider returns a token for the supplied configuration.
+func NewTokenProvider(config Config, stopCh <-chan struct{}) (*TokenProvider, error) {
+	if config.TokenExpiry == 0 {
+		config.TokenExpiry = time.Hour
+	}
+
+	tp := &TokenProvider{
+		config: config,
+	}
+	if _, err := tp.Token(); err != nil {
+		return nil, errors.Wrap(err, "mint token")
+	}
+	go tp.refreshLoop(time.Duration(config.TokenExpiry/10), stopCh)
```

Comment:
How did we come up with the divide by 10 value? If we're starting off with an hour and dividing by 10, this means it will run every 6 minutes but doesn't that also mean if we miss the refresh on the hour mark, we will need to wait another 6 minutes to try and refresh? Wouldn't this mean the zms requests would fail for this period?

Reply:
Setting to 1/10th of the overall expiry, would give at least 2 chances(between 75% - 100% of expiry time) for the refreshLoop to rotate the token. The attempts every 6m would be no-op except within the 15m grace period

Reply:
Got it, that makes sense.